### PR TITLE
 KFSPTS-11821 update AWS billing job to handle multiple default KFS account formats

### PR DESCRIPTION
**Pull request by jhulslander**
This PR updates the cost center and default account translation processing to be consistent, and added functionality for chart*account.

### CODE REVIEW

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java [hunk 1]
```diff
@@ -41,222 +43,268 @@
...
+        line.setSubObjectCode(accountDto.getKfsSubObject());
+        line.setProjectCode(accountDto.getKfsProject());
+        line.setOrgRefId(accountDto.getKfsOrgRefId());
+        line.setLineDescription(StringUtils.EMPTY);
+        return line;
+    }
+    
+    private void fixAccountingLine(AccountingXmlDocumentAccountingLine xmlAccountingLine, AmazonKfsAccountDTO defaultAccountDto) {
+        boolean logErrorMessage = true;
+        fixChart(xmlAccountingLine, defaultAccountDto, logErrorMessage);
```
Comment (db483) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java]: The "fix" method names are a bit confusing. I wonder if "validateAndSetDefaults" would be a good replacement for the "fix" part

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java [hunk 2]
```diff
@@ -41,222 +43,268 @@
...
+        fixAccount(xmlAccountingLine, defaultAccountDto, logErrorMessage);
+        foxSubAccount(xmlAccountingLine, defaultAccountDto, logErrorMessage);
+        foxObjectCode(xmlAccountingLine, defaultAccountDto, logErrorMessage);
```
Comment (db483) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java]: spelling error

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java [hunk 3]
```diff
@@ -41,222 +43,268 @@
...
+        fixSubObjectCode(xmlAccountingLine, defaultAccountDto, logErrorMessage);
+        fixProjectCode(xmlAccountingLine, defaultAccountDto, logErrorMessage);
+        foxOrgRefId(xmlAccountingLine, defaultAccountDto, logErrorMessage);
```
Comment (db483) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java]: spelling error

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java [hunk 4]
```diff
@@ -41,222 +43,268 @@
...
-            xmlAccountingLine = parseCostCenterGroupLevel(defaultAccountString);
-        }
-        else if (StringUtils.countMatches(defaultAccountString, "-") == 1) {
-            xmlAccountingLine.setAccountNumber(StringUtils.split(defaultAccountString, "-")[0]);
-            xmlAccountingLine.setSubAccountNumber(StringUtils.split(defaultAccountString, "-")[1]);
-        }
-        else {
-            xmlAccountingLine.setAccountNumber(defaultAccountString);
+    
+    private void logAccountValidationError(String message, boolean logErrorMessage) {
```
Comment (db483) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java]: Could you help me understand why sometimes we want to log the error message and other times we don't?

File: src/main/java/edu/cornell/kfs/fp/batch/businessobject/AmazonBillResultsDTO.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+public class AmazonBillResultsDTO {
+    public String masterAccountNumber;
+    public String masterAccountName;
+    public boolean successfullyProcessed;
+    public int numberOfAwsAccountInCloudCheckr;
+    public int xmlCreationCount;
+    public List<String> awsAccountWithoutDefaultAccount;
+    public List<String> awsAccountWithExistingDI;
+    public List<String> awsAccountGeneratedDIxml;
+    public List<AmazonKfsAccountDTO> defaultAccountErrors;
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/businessobject/AmazonBillResultsDTO.java]: OK, will change

File: src/main/java/edu/cornell/kfs/fp/batch/businessobject/AmazonBillResultsDTO.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    public List<AmazonKfsAccountDTO> costCenterErrors;
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/businessobject/AmazonBillResultsDTO.java]: OK, will change

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java [hunk 7]
```diff
@@ -41,222 +43,268 @@
...
+        line.setOrgRefId(accountDto.getKfsOrgRefId());
+        line.setLineDescription(StringUtils.EMPTY);
+        return line;
+    }
+    
+    private void fixAccountingLine(AccountingXmlDocumentAccountingLine xmlAccountingLine, AmazonKfsAccountDTO defaultAccountDto) {
+        boolean logErrorMessage = true;
+        fixChart(xmlAccountingLine, defaultAccountDto, logErrorMessage);
+        fixAccount(xmlAccountingLine, defaultAccountDto, logErrorMessage);
+        foxSubAccount(xmlAccountingLine, defaultAccountDto, logErrorMessage);
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java]: will fix

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java (hunk 1 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java]: I will change to use the different naming convention

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java (hunk 2 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java]: I'll change to use validateAndSetDefaults convention

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java (hunk 3 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java]: I'll change to use the validateAndSetDefaults convention

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java (hunk 4 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java]: We run the validation multiple times.  Once to identify default accounts or cost centers that have errors.  The second time is when the actual transaction lines are being built.  We only want to log once so we don't have duplicate log statements.

The validation routines run very quickly and doesn't cause a performance issue running them more than once.

**Comment by jhulslander** [comment]
this is . ready for review again

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java [hunk 8]
```diff
@@ -41,222 +43,268 @@
...
+    }
 
-        if (!StringUtils.equalsIgnoreCase(defaultAccountString, CuFPConstants.AmazonWebServiceBillingConstants.INTERNAL_KFS_ACCOUNT_DESCRIPTION) &&
-                !validateAccount(xmlAccountingLine.getChartCode(), xmlAccountingLine.getAccountNumber())) {
-            xmlAccountingLine = getDefaultAccountingLine(defaultAccountString);
+    private void validateAndSetDefaultsSubAccount(AccountingXmlDocumentAccountingLine xmlAccountingLine,
+            AmazonKfsAccountDTO defaultAccountDto, boolean logErrorMessage) {
+        if (StringUtils.isNotBlank(xmlAccountingLine.getSubAccountNumber())
+                && !validateSubAccountNumber(xmlAccountingLine.getChartCode(), xmlAccountingLine.getAccountNumber(), xmlAccountingLine.getSubAccountNumber(), logErrorMessage)) {
+            if (validateSubAccountNumber(xmlAccountingLine.getChartCode(), xmlAccountingLine.getAccountNumber(), defaultAccountDto.getKfsSubAccount(), logErrorMessage)) {
```
Comment (db483) [inline src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java]: what happens if default sub Account is invalid?

**Comment by jhulslander** [comment]
Daniella and I talked about this.  chart, account and object are required so we don't validate the default values of them, the other fields aren't required so if the default value isn't valid, we can ignore it.  In going over this I found a bug, which I will correct, and we also talked about how to make this clearer

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java (hunk 8 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/AwsAccountingXmlDocumentAccountingLineServiceImpl.java]: it gets set to an empty string

**Comment by jhulslander** [comment]
I've made  some changes, this is ready for review again

**Pull request merged**